connector/connect/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamHandler.scala

```diff
@@ -117,7 +126,70 @@ class SparkConnectStreamHandler(responseObserver: StreamObserver[Response]) exte
       responseObserver.onNext(response.build())
     }
 
-    responseObserver.onNext(sendMetricsToResponse(clientId, rows))
+    responseObserver.onNext(sendMetricsToResponse(clientId, dataframe))
+    responseObserver.onCompleted()
+  }
+
+  def processRowsAsArrowBatches(clientId: String, dataframe: DataFrame): Unit = {
+    val spark = dataframe.sparkSession
+    val schema = dataframe.schema
+    // TODO: control the batch size instead of max records
+    val maxRecordsPerBatch = spark.sessionState.conf.arrowMaxRecordsPerBatch
+    val timeZoneId = spark.sessionState.conf.sessionLocalTimeZone
+
+    val rows = dataframe.queryExecution.executedPlan.execute()
+    var numBatches = 0L
+
+    if (rows.getNumPartitions > 0) {
+      val batches = rows.mapPartitionsInternal { iter =>
+        ArrowConverters
+          .toArrowBatchIterator(iter, schema, maxRecordsPerBatch, timeZoneId)
+      }
+
+      val obj = new Object
+
+      val processPartition = (iter: Iterator[(Array[Byte], Long, Long)]) => iter.toArray
+
+      val resultHandler = (partitionId: Int, taskResult: Array[(Array[Byte], Long, Long)]) =>
+        obj.synchronized {
+          var batchId = partitionId.toLong << 33
+          taskResult.foreach { case (bytes, count, size) =>
+            val response = proto.Response.newBuilder().setClientId(clientId)
+            val batch = proto.Response.ArrowBatch
+              .newBuilder()
+              .setBatchId(batchId)
+              .setRowCount(count)
+              .setUncompressedBytes(size)
+              .setCompressedBytes(bytes.length)
+              .setData(ByteString.copyFrom(bytes))
+              .build()
+            response.setArrowBatch(batch)
+            responseObserver.onNext(response.build())
```

Comment:
This callback is currently executed by the DAGScheduler thread, and if this expensive no job/stage can be scheduler during this call. We really should move this to a separate thread.

Reply:
ok will update